### PR TITLE
gh-104976: Ensure trailing dedent tokens are emitted as the previous tokenizer

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-104976

### CODE REVIEW

**Review by rllane17 — APPROVED** [review]

**Comment by lysnikolaou** [comment]
@pablogsal Can you provide a very quick explanation of how this works and changes the location of the DEDENT token? Nothing too detailed, just a quick overview of the though process, cause I'm having difficulties understanding this.

**Comment by pablogsal** [comment]
> Can you provide a very quick explanation of how this works and changes the location of the DEDENT token?

Yep: the objective is that inplicit dedents like these:
```
if X:
  if Y:
     ...
```

are reported in the same line as the `ENDMARKER`. For doing this we note that trailing dedent tokens are emitted in a situation where the tokenizer has reached `EOF` so `tok->done` is `E_EOF` for all of them, which is great because this allows to identify them easily. We add now a branch for these to put the line number that is expected as the previous tokenizer.

Unfortunately, this forces us to handle `ENDMARKER` in the extension because otherwise we don't have enough information to get correctly the line for it because checking the last token may now have these "artificial" new lines.

**Comment by pablogsal** [comment]
> @pablogsal Can you provide a very quick explanation of how this works and changes the location of the DEDENT token? Nothing too detailed, just a quick overview of the though process, cause I'm having difficulties understanding this.

Ah wait, I forgot to push some commits 😓

File: Lib/test/test_tokenize.py [hunk 1]
```diff
@@ -1174,7 +1174,7 @@ def readline():
 
         # skip the initial encoding token and the end tokens
         tokens = list(_tokenize(readline(), encoding='utf-8'))[:-2]
-        expected_tokens = [TokenInfo(3, '"ЉЊЈЁЂ"', (1, 0), (1, 7), '"ЉЊЈЁЂ"')]
+        expected_tokens = [TokenInfo(3, '"ЉЊЈЁЂ"', (1, 0), (1, 7), '"ЉЊЈЁЂ"\n')]
```
Comment (pablogsal) [inline Lib/test/test_tokenize.py]:
```suggestion
        expected_tokens = [TokenInfo(3, '"ЉЊЈЁЂ"', (1, 0), (1, 7), '"ЉЊЈЁЂ"')]
```

**Review by lysnikolaou — APPROVED** [review]
Yup! Looks good!

**Pull request merged**